### PR TITLE
feat: add controller to periodically run Helm Chart Tests

### LINKED ISSUE #106: [FEAT] - optionally run ChartTests for a Plugin
### Priority

None

### Description

Currently the HelmController only checks if a Plugin's Release could be installed/upgraded successfully. However, it there are no additional checks to verify that the Helm Release is working as expected.
With [Chart Tests](https://helm.sh/docs/topics/chart_tests/) it is possible to run a set of tests after a Helm action has been performed to validate that the deployed components work as intended.

> Example tests:
> 
> Validate that your configuration from the values.yaml file was properly injected.
> Make sure your username and password work correctly
> Make sure an incorrect username and password does not work
> Assert that your services are up and correctly load balancing
> etc.

Acceptance Criteria

- Plugin Developer can provide ChartTests as part of the _PluginDefinition_'s Helm Chart.
- ChartTests are run only if the Release is deployed, and helm hooks with `"helm.sh/hook": test` exist (see https://github.com/cloudoperators/greenhouse/blob/main/pkg/helm/helm.go#L53)
- ChartTests are run if the Plugin changes and on a configurable interval 
- Results of the ChartTests are reflected in the _Plugins_ statusConditions as `type: ChartTestResult`, defaulted in helmController
- Failure of these tests should be also be reflected in prometheus counter metric
   `greenhouse_plugin_charttest_runs_total{cluster="", plugin="", namespace="", result="Success|Error|NoTests"}`
- Integration test are in place 

### Status 
```yaml
statusConditions:
  conditions:
    - lastTransitionTime: "2024-08-01T11:21:58Z"
      status: "Unknown"
      type: NoHelmChartTestFailures
    - lastTransitionTime: "2024-08-01T11:16:58Z"
      status: "True"
      type: NoHelmChartTestFailures
      message: "Success"
    - lastTransitionTime: "2024-08-01T11:21:58Z"
      status: "False"
      type: NoHelmChartTestFailures
      message: "Failed: <error message>"
```
_No response_

### PR DESCRIPTION
**Pull request by ibakshay**
### **Description**

This pull request introduces support for periodically running Helm ChartTests from a controller. 

Currently, the HelmController only verifies whether a Plugin's Release can be installed/upgraded successfully. There are no additional checks to ensure that the Helm Release functions as expected. 

By periodically running Helm Chart Tests from a controller, it's possible to execute a set of native Helm Chart integration tests to validate that the deployed components work correctly. 

### **Impact on Plugin Status**

1. **When Chart Test is Successful**

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: Plugin
metadata:
  name: kube-monitoring-obs-eu-de-1
  namespace: test-org
spec:
  # ... (rest of your spec)
status:
# ... (rest of your status)
  statusConditions:
    conditions:
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      message: ready
      status: "True"
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      message: Helm Chart Test is successful
      status: "True"
      type: HelmChartTestResult
```

2. **When the plugin is not Deployed Successfully**

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: Plugin
metadata:
  name: kube-monitoring-obs-eu-de-1
  namespace: test-org
spec:
  # ... (rest of your spec)
status:
# ... (rest of your status)
  statusConditions:
    conditions:
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      message: ready
      status: "True"
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      status: Unknown
      type: HelmChartTestResult
```

3. **When there are no Helm Chart Tests for the Plugin**

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: Plugin
metadata:
  name: kube-monitoring-obs-eu-de-1
  namespace: test-org
spec:
  # ... (rest of your spec)
status:
# ... (rest of your status)
  statusConditions:
    conditions:
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      message: ready
      status: "True"
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      status: "False"
      message: "Helm Chart Test is not defined"
      type: HelmChartTestResult
```

4. **When the Helm Chart Test Fails**

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: Plugin
metadata:
  name: kube-monitoring-obs-eu-de-1
  namespace: test-org
spec:
  # ... (rest of your spec)
status:
# ... (rest of your status)
  statusConditions:
    conditions:
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      message: ready
      status: "True"
    - lastTransitionTime: "2024-08-12T13:06:44Z"
      status: "False"
      message: 'Helm Chart Test failed: 1 error occurred:  pod kube-monitoring-obs-eu-de-1-test failed. Please check the logs of the failed test pod for more information'
      type: HelmChartTestResult
```

fixes #106 
depends on https://github.com/cloudoperators/greenhouse/pull/398

### CODE REVIEW

File: pkg/helm/helm.go [hunk 1]
```diff
@@ -93,6 +93,34 @@ func InstallOrUpgradeHelmChartFromPlugin(ctx context.Context, local client.Clien
...
+	if results.Hooks != nil {
+	outer:
+		for _, hook := range results.Hooks {
+			for _, event := range hook.Events {
+				if event == release.HookTest {
+					hasTestHook = true
+					break outer // Break out of both loops when a test hook is found
+				}
+			}
+		}
```
Comment (IvoGoman) [inline pkg/helm/helm.go:118]: NIT: What do you think about using the _slices_ package instead?

The `slices.ContainsFunc(..)` returns on the first hit, also the `slices.Contains(..)` returns on the first hit.

```suggestion
		hasTestHook = slices.ContainsFunc(results.Hooks, func(h *release.Hook) bool {
			return slices.Contains(h.Events, release.HookTest)
		})
```

**Review by IvoGoman — COMMENTED** [review]
Looks good, besides adding the additional metric.

**Comment by ibakshay** [comment]
> Looks good, besides adding the additional metric.

great suggestion to add metrics for chart_test_counts. I have added in this commit https://github.com/cloudoperators/greenhouse/pull/441/commits/6a5f3655184adf50a461f0023bd172efd633a258

// cc: @richardtief

File: pkg/helm/helm.go (hunk 1 above)
Comment (ibakshay) [reply pkg/helm/helm.go]: thank you for the suggestion. Implemented in this https://github.com/cloudoperators/greenhouse/pull/441 commit.

**Comment by ibakshay** [comment]
Tested the created  metrics `greenhouse_plugin_chart_test_runs_total` locally and it works as expected 🚀
```shell
curl localhost:8080/metrics
### Response ####
----
----
greenhouse_plugin_chart_test_runs_total{cluster="remote-cluster-1-control-plane",namespace="test-org",plugin="kube-monitoring-obs-eu-de-1",result="Error"} 5
greenhouse_plugin_chart_test_runs_total{cluster="remote-cluster-1-control-plane",namespace="test-org",plugin="kube-monitoring-obs-eu-de-1",result="Success"} 1
greenhouse_plugin_chart_test_runs_total{cluster="remote-cluster-1-control-plane",namespace="test-org",plugin="thanos-test",result="NoTests"} 2
```

**Review by IvoGoman — APPROVED** [review]

File: pkg/controllers/plugin/helm_chart_test_controller.go [hunk 2]
```diff
@@ -95,15 +115,24 @@ func (r *HelmChartTestReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		prometheusLabels["result"] = "Error"
+		chartTestRunsTotal.With(prometheusLabels).Inc()
+
 		return ctrl.Result{}, err
 	}
 
 	if !hasHelmChartTest {
 		noHelmChartTestFailuresCondition.Status = metav1.ConditionTrue
-		noHelmChartTestFailuresCondition.Message = "Helm Chart Test is not defined"
+		noHelmChartTestFailuresCondition.Message = "No Helm Chart Test found. It is strongly recommended to have Helm Chart Tests for your Helm Chart of the plugin."
```
[the remark this answers is not in the document]
Comment (IvoGoman) [reply pkg/controllers/plugin/helm_chart_test_controller.go:126]: What do you think about this: `No Helm Chart Tests defined by the PluginDefinition`

**Review by IvoGoman — APPROVED** [review]

**Pull request merged**